fluentinterface/src/main/java/com/iluwatar/fluentinterface/fluentiterable/lazy/LazyFluentIterable.java

```diff
@@ -0,0 +1,236 @@
+                            }
+
+                            return null;
+                        }
+                    }
+                };
+            }
+        };
+    }
+
+    /**
+     * Uses the Iterable interface's forEach method to apply a given function
+     * for each object of the iterator. Is a terminating operation.
+     * @param action the action for each object
+     */
+    @Override
+    public void forEachDo(Consumer<? super TYPE> action) {
+        Iterator<TYPE> newIterator = iterable.iterator();
+        while(newIterator.hasNext()) {
+            action.accept(newIterator.next());
+        }
+    }
+
+    /**
+     * Can be used to collect objects from the iteration. Is a terminating operation.
+     * @return an option of the first object of the iteration
+     */
+    @Override
+    public Optional<TYPE> first() {
+        Optional result = Optional.empty();
+        List<TYPE> list = first(1).asList();
+        if(!list.isEmpty()) {
+            result = Optional.of(list.get(0));
+        }
+
+        return result;
+    }
+
+    /**
+     * Can be used to collect objects from the iteration. Is a terminating operation.
+     * @param count defines the number of objects to return
+     * @return the same FluentIterable with a collection decimated to a maximum of 'count' first objects.
+     */
+    @Override
+    public FluentIterable<TYPE> first(int count) {
+        return new LazyFluentIterable<TYPE>() {
+            @Override
+            public Iterator<TYPE> iterator() {
+                return new DecoratingIterator<TYPE>(iterable.iterator()) {
+                    int currentIndex = 0;
+
+                    @Override
+                    public TYPE computeNext() {
+                        if(currentIndex < count) {
+                            if(fromIterator.hasNext()) {
+                                TYPE candidate = fromIterator.next();
+                                currentIndex++;
+                                return candidate;
+                            }
+                        }
+                        return null;
+                    }
+                };
+            }
+        };
+    }
+
+    /**
+     * Can be used to collect objects from the iteration. Is a terminating operation.
+     * @return an option of the last object of the iteration
+     */
+    @Override
+    public Optional<TYPE> last() {
+        Optional result = Optional.empty();
+        List<TYPE> list = last(1).asList();
+        if(!list.isEmpty()) {
+            result = Optional.of(list.get(0));
+        }
+
+        return result;
+    }
+
+    /**
+     * Can be used to collect objects from the iteration. Is a terminating operation.
+     * @param count defines the number of objects to return
+     * @return the same FluentIterable with a collection decimated to a maximum of 'count' last objects
+     */
+    @Override
+    public FluentIterable<TYPE> last(int count) {return new LazyFluentIterable<TYPE>() {
```

Comment:
I think it would be nice to have it, because that's probably one thing people have problems to implement by themselves - regardless if the operation would be cheap or expensive. I added additional documentation instead.